File: Wordfolio.Frontend/src/features/collections/api/mappers.ts
import {
    CollectionWithVocabularyCountResponse,
    CollectionResponse,
    CreateCollectionRequest,
    UpdateCollectionRequest,
    VocabularyWithEntryCountResponse,
} from "./collectionsApi";
import {
    Collection,
    CollectionWithVocabularyCount,
    VocabularyWithEntryCount,
} from "../types";
import { CollectionFormData } from "../schemas/collectionSchemas";

export const mapCollectionDetail = (
    response: CollectionResponse
): Collection => ({
    id: response.id,
    name: response.name,
    description: response.description,
    createdAt: new Date(response.createdAt),
    updatedAt: response.updatedAt ? new Date(response.updatedAt) : null,
});

export const mapVocabularyWithEntryCount = (
    collectionId: number,
    response: VocabularyWithEntryCountResponse
): VocabularyWithEntryCount => ({
    id: response.id,
    collectionId,
    name: response.name,
    description: response.description,
    entryCount: response.entryCount,
    createdAt: new Date(response.createdAt),
    updatedAt: response.updatedAt ? new Date(response.updatedAt) : null,
});

export const mapCollectionWithVocabularyCount = (
    response: CollectionWithVocabularyCountResponse
): CollectionWithVocabularyCount => ({
    id: response.id,
    name: response.name,
    description: response.description,
    vocabularyCount: response.vocabularyCount,
    createdAt: new Date(response.createdAt),
    updatedAt: response.updatedAt ? new Date(response.updatedAt) : null,
});

export const mapToCreateCollectionRequest = (
    data: CollectionFormData
): CreateCollectionRequest => ({
    name: data.name,
    description: data.description,
});

export const mapToUpdateCollectionRequest = (
    data: CollectionFormData
): UpdateCollectionRequest => ({
    name: data.name,
    description: data.description,
});

File: Wordfolio.Api/Wordfolio.Api/Api/CollectionsHierarchy/Mappers.fs
module Wordfolio.Api.Api.CollectionsHierarchy.Mappers

open Wordfolio.Api.Api.CollectionsHierarchy.Types
open Wordfolio.Api.Domain
open Wordfolio.Api.Domain.CollectionsHierarchy
open Wordfolio.Api.Infrastructure.ResourceIdEncoder

let toVocabularyWithEntryCountResponse
    (encoder: IResourceIdEncoder)
    (vocabulary: VocabularyWithEntryCount)
    : VocabularyWithEntryCountResponse =
    { Id = encoder.Encode(VocabularyId.value vocabulary.Id)
      Name = vocabulary.Name
      Description = vocabulary.Description
      CreatedAt = vocabulary.CreatedAt
      UpdatedAt = vocabulary.UpdatedAt
      EntryCount = vocabulary.EntryCount }

let toCollectionWithVocabulariesResponse
    (encoder: IResourceIdEncoder)
    (collection: CollectionWithVocabularies)
    : CollectionWithVocabulariesResponse =
    { Id = encoder.Encode(CollectionId.value collection.Id)
      Name = collection.Name
      Description = collection.Description
      CreatedAt = collection.CreatedAt
      UpdatedAt = collection.UpdatedAt
      Vocabularies =
        collection.Vocabularies
        |> List.map(toVocabularyWithEntryCountResponse encoder) }

let toCollectionWithVocabularyCountResponse
    (encoder: IResourceIdEncoder)
    (collection: CollectionWithVocabularyCount)
    : CollectionWithVocabularyCountResponse =
    { Id = encoder.Encode(CollectionId.value collection.Id)
      Name = collection.Name
      Description = collection.Description
      CreatedAt = collection.CreatedAt
      UpdatedAt = collection.UpdatedAt
      VocabularyCount = collection.VocabularyCount }
